agents/timestream_aggregator/timestream_aggregator.py

```diff
@@ -114,40 +410,19 @@ def start_aggregation(self, session, params=None):
 
         self.log.info("Data directory set to {}".format(self.data_dir))
         self.log.info("New file every {} seconds".format(self.time_per_file))
-        self.log.info("Listening to {}:{}".format(self.address, self.port))
+        self.log.info("Listening to {}".format(self.address))
 
-        reader = core.G3Reader("tcp://{}:{}".format(self.address,
-                                                    self.port))
-        writer = None
-
-        last_meta = None
         self.is_streaming = True
 
+        recorder = FrameRecorder(self.time_per_file, self.address,
+                                 self.data_dir, self.log)
+
         while self.is_streaming:
-            if writer is None:
-                start_time = time.time()
-                filepath = _create_file_path(start_time, self.data_dir)
-                self.log.info("Writing to file {}".format(filepath))
-                writer = core.G3Writer(filename=filepath)
-
-                if last_meta is not None:
-                    writer(last_meta)
-
-            frames = reader.Process(None)
-            for f in frames:
-                if f.type == core.G3FrameType.Observation:
-                    last_meta = f
-                writer(f)
-                writer.Flush()
-
-            if (time.time() - start_time) > self.time_per_file:
-                writer(core.G3Frame(core.G3FrameType.EndProcessing))
-                writer = None
+            recorder.run()
 
         # Once we're done writing close out the file
-        if writer is not None:
-            writer(core.G3Frame(core.G3FrameType.EndProcessing))
-            writer = None
+        if recorder.writer is not None:
+            recorder.close_file()
```

Comment:
Instead of this "invasive" check/call that needs to know internal rules about FrameRecorder, consider defining a FrameRecorder.finalize() function that, by definition, does all the necessary things to properly terminate a run.  Alternately you could put this code into FrameRecorder.__del__ (the destructor), and just do "del recorder" to clean things up explicitly (though the same should happen when recorder goes out of scope, including if there's some kind of exception).

Reply:
I like the destructor suggestion, so I did that and now explicitly call `del recorder`.